src/System.Memory/src/System/SpanExtensions.cs

```diff
@@ -274,5 +274,104 @@ public static void CopyTo<T>(this T[] array, Span<T> destination)
         {
             new ReadOnlySpan<T>(array).CopyTo(destination);
         }
+
+        /// <summary>
+        /// Determines whether two sequences overlap.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second)
+        {
+            return Overlaps((ReadOnlySpan<T>)first, second);
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap and outputs the element offset.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            return Overlaps((ReadOnlySpan<T>)first, second, out elementOffset);
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap.
+        /// </summary>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second)
+        {
+            // note: the calculations below overflow if second.IsEmpty;
+            //       the case second.IsEmpty MUST be handled separately
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                return (uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()) ||
+                       (uint)byteOffset > (uint)-(second.Length * Unsafe.SizeOf<T>());
```

Comment:
@ektrah the problem is the `(uint)byteOffset > (uint)-(second.Length * Unsafe.SizeOf<T>())` check when `second.Length == 0` right?

Reply:
Yes, exactly.

Reply:
I actually just thought of simply overflowing the end and checking against the start of second:
```csharp
(uint)byteOffset + second.Length * Unsafe.SizeOf<T>()) < (uint)byteOffset
```